servers/src/main/java/tachyon/master/file/FileSystemMaster.java

```diff
@@ -1340,33 +1415,35 @@ boolean unmountInternal(TachyonURI tachyonPath) throws InvalidPathException {
     return mMountTable.delete(tachyonPath);
   }
 
-
   /**
    * MasterInodeTTL periodic check.
    */
-  public final class MasterInodeTTLCheckExecutor implements HeartbeatExecutor {
-    // TODO: current implementation needs to be improved by using a more efficient datastructure
-    // such as hourly bucketized inodefilelist
+  private final class MasterInodeTTLCheckExecutor implements HeartbeatExecutor {
     @Override
     public void heartbeat()  {
       synchronized (mInodeTree) {
-        List<Inode> inodes = mInodeTree.getInodeChildrenRecursive(mInodeTree.getRoot());
-        for (Inode inode : inodes) {
-          if (inode.isFile() && !inode.isPinned()) {
-            InodeFile iFile = (InodeFile) inode;
-            long ttl = iFile.getTTL();
-            if (ttl > 0 && System.currentTimeMillis() - iFile.getCreationTimeMs() > ttl * 1000) {
+        if (mTTLBuckets.size() > 0 && mTTLBuckets.get(0).getTTLIntervalStartTimeMs() < 0) {
+          // The ttl interval start of this bucket is smaller than 0, so all files in this bucket
+          // should no longer live.
+          for (InodeFile file : mTTLBuckets.get(0).getFiles()) {
+            if (!file.isDeleted() && !file.isPinned()) {
```

Comment:
For simplicity, I think TTLChecker should only care about ttl, no worry about whether it's pinned, otherwise, we need to define the behavior when ttl has been expired and the file is unpinned, I'll document that ttl will overwrite pin on the configuration documentation.